src/output.rs

```diff
@@ -63,6 +70,11 @@ fn print_entry_colorized(path: &Path, config: &FdOptions, ls_colors: &LsColors)
             // Everything else uses a separator that is painted the same way as the component.
             _ => style.paint(path::MAIN_SEPARATOR.to_string()).to_string(),
         };
+
+        if wants_to_quit.load(Ordering::Relaxed) {
+            write!(handle, "\n")?;
+            process::exit(0);
```

Comment:
Sorry to complain, since I likely won't have the time to contribute a fix myself, but I just noticed this while looking through the change log. According to [the docs](https://doc.rust-lang.org/std/process/fn.exit.html), it's generally not good practice to just `process::exit` in the middle of the program's execution.

> Note that because this function never returns, and that it terminates the process, no destructors on the current stack or any other thread's stack will be run. If a clean shutdown is needed it is recommended to only call this function at a known point where there are no more destructors left to run.

That is, it should really be run at the very end of the main function. So rather than call `process::exit` here, it should return an error value of some kind that gets propagated all the way up to the main function.

Also, if you're going to handle signals, it's probably better to handle them in the general case, rather than just when there are colors. Messing up the terminal colors is obviously the more pressing case because it messes up the user's whole shell, but the program as a whole should handle exiting gracefully.

Also, this should probably not exit with code 0, which generally means success. Ctrl-C is usually an error condition for a program.

Reply:
@dead10ck thanks for bringing this to our attention.

I was not aware of this behavior from `process::exit`, a quick fix to remove the `process:exit(0)` would be to replace it with something like `return Err(Error::new(ErrorKind::Interrupted, "Interrupted"));`. I'm just not sure if we need to do something similar in the other spawned thread though. What happens if a rust program exits with outstanding threads executing?

We could handle the terminal signal in the general case, but we handle SIGINT for tear down purposes. We don't have to tear down in the non-colorized case so there's no point in handling SIGINT at that point. We can just use the default SIGINT handling to exit when output is not colorized.

`SIGINT` is generally for graceful shutdown of a program, I would not say it is generally an error condition. The application handles `SIGINT` and exits in a known and expected fashion, this would be a successful execution of the program in my opinion.

Reply:
> I was not aware of this behavior from process::exit, a quick fix to remove the process:exit(0) would be to replace it with something like return Err(Error::new(ErrorKind::Interrupted, "Interrupted"));. I'm just not sure if we need to do something similar in the other spawned thread though. What happens if a rust program exits with outstanding threads executing?

I'm sorry, I don't actually have much experience with this in Rust; this is just a tidbit of information I happened to recall when glancing over the changes.

> We could handle the terminal signal in the general case, but we handle SIGINT for tear down purposes. We don't have to tear down in the non-colorized case so there's no point in handling SIGINT at that point. We can just use the default SIGINT handling to exit when output is not colorized.

I haven't done my research here, but if I'm not mistaken, the default behavior is not graceful—it essentially just aborts the program in the middle, just like calling `process::exit`. In any case, though, I think if you're going to go to the trouble of handling graceful exiting, why not handle it everywhere, if it's not too much extra effort?

> SIGINT is generally for graceful shutdown of a program, I would not say it is generally an error condition. The application handles SIGINT and exits in a known and expected fashion, this would be a successful execution of the program in my opinion.

I can understand your perspective, but this is not conventional. [POSIX specifies that a process terminated by a signal should have an exit code greater than 128.](https://unix.stackexchange.com/a/99134/152675) You can test this for yourself with any other standard Linux utility.

Reply:
Moreover, it occurred to me that if you only handle the signal when there are colors, *and* exit with 0 in that case, `fd` will exit with 0 when there are colors and non-0 when there aren't:

```zsh
➜  ~ fd '' / --color always
...
/proc/9/coredump_filter
^C
/
➜  ~ echo $?               
0
➜  ~ fd '' / --color never
...
/var/log/cups/page_log.3
/var/log/puppetlabs/puppet
/var/log/cups/access_log.4
/var/log/cups/access_log.2
/var/log/cups/page_log.1
/var/log/cu^C
➜  ~ echo $?
130
```

Reply:
@dead10ck, looks like you're right about the exit code.

It should also be noted that while `process::exit` does not call destructors on exit the memory is not going to really be leaked on any standard unix or windows operating system. The OS keeps track of all memory allocated by a program and will free that memory when the program exits. Only on specialized operating systems such as real-time operating systems do you run the chance of the OS not cleaning up memory after program exit.

For the time being I think we can just replace the exit code with 130 so fd reports it was killed  by SIGINT correctly.

On the topic of general SIGINT handling, we have only one special case that requires delaying the exit until after the thread is done writing to the terminal. There's no need in delaying exit when terminal output is not colorized, hence custom shutdown only in the colorized case.

@sharkdp do you have any preferences on these matters?

Reply:
@dead10ck Thank you for pointing this out

@Doxterpepper In this case, exiting with 130 seems like a reasonable first step. However, we should maybe open a new ticket in order to refactor the code such that error handling like this would be easier.